File: 프로그래머스/lv2/42577. 전화번호 목록/전화번호 목록.py
def solution(phone_book):
    prefix_dict = {}
    answer = True
    for phone_num in phone_book:
        prefix_dict[phone_num] = 1
        
    for phone_num in phone_book:
        prefix = ""
        for num in phone_num:
            prefix += num
            if prefix in prefix_dict and prefix != phone_num:
                return False
    return answer

File: 프로그래머스/2/42577. 전화번호 목록/전화번호 목록.java
import java.util.*;

class Solution {
    
    Map<String, Boolean> phoneBook = new HashMap<>();
    
    
    public boolean solution(String[] phone_book) {
        
        for(String number: phone_book){
            phoneBook.put(number, true);
        }
        
        for(String number: phone_book){
    
            StringBuilder prefix = new StringBuilder();
            for(Character num: number.toCharArray()){
                prefix.append(num);
                if(phoneBook.containsKey(prefix.toString()) && !prefix.toString().equals(number)){
                    return false;
                }
            }
             
        }
        return true;
    }
}